README.md
# Facebook Auto "Most Recent" Stories

This script is a Userscript to work with Facebook.

With this, your feed is automatically sorted to the most recent activity.

As well, if you like to see the most recent posts in the groups, this script do the job for you.

> Note: You must install this userscript from here: [Greasy Fork: Facebook \"Most Recent\" Stories](https://greasyfork.org/en/scripts/382099-facebook-auto-most-recent-stories "Facebook \"Most Recent\" Stories").

## How to report an issue?
I use the GitHub as bug tracker. If you find a bug, go to [here](https://github.com/Mettafox/Facebook-Userscript/issues "Facebook-Userscript bug tracker") and click on New issue.

Do not submit a issue solely saying that the script has a bug or something is not working. Give me as much detail as possible and how to reproduce the problem. If you do not give me details I do not have the ability to guess what it is or what is causing the problem. For more informations how to report an issue, see [this bug report template](https://github.com/Mettafox/Facebook-Userscript/blob/master/bug_report.md "bug_report.md").

If you know how to solve the problem and want to contribute, please open a Pull Request with your solution.

## Group feed: ## 
If you want to see **New Activity** instead of **Recent Posts**, you need to change that in the script.   
To do so, you need to change the `gpop[*]` index from `gpop[0]` to `gpop[1]` everywhere this variable is used.


CHANGELOG.md
Changelog
======

## [v0.1559010793 (28/05/2019)](https://greasyfork.org/en/scripts/382099-facebook-auto-most-recent-stories) - Minor Update ##

### Updated ###
* An extra verification was added to the **"gdisc"** click event to avoid an unwanted event when click in the About tab (pages).   
   > The nav menu in the pages, uses the same class names as the nav menu in the groups, so, if you click in the About tab, the "click" event will be triggered, just like when you click in the discussion tab.  
To avoid that, an extra verification was added.

## [v0.1558138696 (18/05/2019)](https://greasyfork.org/en/scripts/382099-facebook-auto-most-recent-stories) ##

### New ###
* Now when you click in the groups links in the Shortcut menu, you will be redirected to that group but with the option CHRONOLOGICAL active in the URL.  
  In this way, the group feed will automaticly change to the most recent publications.  
  **Note:** If you want to change the group feed option, see [Group feed section in the README](https://github.com/Mettafox/Facebook-Userscript/blob/master/README.md#group-feed).

* A 'Click' event was added to the post / comment timestamp. In this way, when you click in the post / comment timestamp, the browser will redirect you to that post / comment and active another 'click' event added to the Discussion button.
  When you click in the Discussion button, you will be redirected to the group with CHRONOLOGICAL option added to the URL.

* If the group URL have this references: **ref=group_header**, **ref=bookmarks** or **ref=direct**, the page will be redirected to the group main page with CHRONOLOGICAL added to the URL.

### Changed ###
* In the previous version, the script perform a modification in the href attribute of the **FB logo** and the **Home button** by adding the Facebook URL with the **?sk=h_chr option**.  
  In this version, I removed that and added a **'click' event** to those two "buttons" to change the URL.  

* The variables are now better organized.

* All code was changed to meet ECMAScript 6 (ECMA-262 6th Edition) standards.  
  * Functions was changed to ***arrow functions***:  
    > An **arrow function expression** is a syntactically compact alternative to a regular function expression, although without its own bindings to the **this**, **arguments**, **super**, or **new.target** keywords. Arrow function expressions are ill suited as methods, and they cannot be used as constructors.  
  * **indexOf()** method replaced by **includes()** method.
* I added some comments to the code for a better understanding of what this or that line do.


main.js
var cur = new URL(window.location), fd = cur["origin"] + "?sk=h_chr", grp = document.querySelectorAll("[data-type='type_group'] > a"), elm = document.querySelectorAll("[data-click='bluebar_logo'] > a, ._3qcu > a"), idx = ["?sk=h_nor", "?ref=logo", "?sk=nf", "?ref=tn_tnmn"], grid = document.querySelectorAll("[property='al:android:url']"), spurl = splitURL(cur["href"], "/"), gdisc = document.querySelectorAll("._2yaa > a"), gpop = ["CHRONOLOGICAL", "RECENT_ACTIVITY"];",
if(cur["href"] === cur["origin"] + "/"){
   window.location.replace(fd);
}
idx.forEach((a) => {
   if(cur["search"].includes(a, 0)){
      window.location.replace(fd);
   }
});
elm.forEach((a) => {
   a.setAttribute("href", fd);
});

//For groups
//Shift the index of the variable gpop[*] (0 = chronological or 1 = recent_activity) to change the group feed disposition.
grp.forEach((a) => {
   a.getAttribute("href");
   var strr = groupSec(a, gpop[0]);
   a.setAttribute("href", strr);",
});
if(cur["href"].includes("groups") && !cur["href"].includes("permalink")){
   if(spurl[5] === "" || spurl[5] === null || spurl[5] === "?ref=group_header" || spurl[5] === "?ref=direct" || spurl[5] !== "?sorting_setting=" + gpop[0]){",
      window.location.replace(sortGrp(grid, "content", gpop[0], cur["origin"] + "/"));
   }
}
window.onload = () => {
   if(gdisc[1] !== undefined){
      gdisc[1].setAttribute("href", sortGrp(grid, "content", gpop[0], cur["origin"] + "/"));
   }
}
if(gdisc[1] !== undefined){
   gdisc[1].setAttribute("href", sortGrp(grid, "content", gpop[0], cur["origin"] + "/"));",
}

//Required functions
//This script requires fbmiscs.js


facebook_most_recent.user.js
// ==UserScript==
// @name        Facebook Auto "Most Recent" Stories
// @version     0.1557115315
// @author      Hélder Ferreira
// @namespace   https://greasyfork.org/users/89591
// @homepageURL https://greasyfork.org/en/scripts/382099-facebook-auto-most-recent-stories
// @description Change Facebook feed and group posts to "Most Recent" Stories / Publications
// @compatible  chrome
// @compatible  firefox
// @compatible  opera
// @compatible  safari
// @compatible  msedge
// @license     MIT License (https://opensource.org/licenses/MIT)
// @icon        http://i.imgur.com/WpjyLA4.png
// @match       https://*.facebook.com/*
// @grant       none
// ==/UserScript==

var cur = new URL(window.location), fd = cur["origin"] + "?sk=h_chr", grp = document.querySelectorAll("[data-type='type_group'] > a"), elm = document.querySelectorAll("[data-click='bluebar_logo'] > a, ._3qcu > a"), idx = ["?sk=h_nor", "?ref=logo", "?sk=nf", "?ref=tn_tnmn"], grid = document.querySelectorAll("[property='al:android:url']"), spurl = splitURL(cur["href"], "/"), gdisc = document.querySelectorAll("._2yaa > a"), gpop = ["CHRONOLOGICAL", "RECENT_ACTIVITY"];",
if(cur["href"] === cur["origin"] + "/"){
   window.location.replace(fd);
}
idx.forEach((a) => {
   if(cur["search"].includes(a, 0)){
      window.location.replace(fd);
   }
});
elm.forEach((a) => {
   a.setAttribute("href", fd);
});

//For groups
//Shift the index of the variable gpop[*] (0 = chronological or 1 = recent_activity) to change the group feed disposition.
grp.forEach((a) => {
   a.getAttribute("href");
   var strr = groupSec(a, gpop[0]);
   a.setAttribute("href", strr);",
});
if(cur["href"].includes("groups") && !cur["href"].includes("permalink")){
   if(spurl[5] === "" || spurl[5] === null || spurl[5] === "?ref=group_header" || spurl[5] === "?ref=direct" || spurl[5] !== "?sorting_setting=" + gpop[0]){",
      window.location.replace(sortGrp(grid, "content", gpop[0], cur["origin"] + "/"));
   }
}
window.onload = () => {
   if(gdisc[1] !== undefined){
      gdisc[1].setAttribute("href", sortGrp(grid, "content", gpop[0], cur["origin"] + "/"));
   }
}
if(gdisc[1] !== undefined){
   gdisc[1].setAttribute("href", sortGrp(grid, "content", gpop[0], cur["origin"] + "/"));",
}

//Required functions
//This script requires fbmiscs.js


main.user.js
// ==UserScript==
// @name        Facebook Auto "Most Recent" Stories
// @version     0.1559010793
// @author      Hélder Ferreira
// @namespace   https://greasyfork.org/users/89591
// @homepageURL https://greasyfork.org/en/scripts/382099-facebook-auto-most-recent-stories
// @description Change Facebook and groups feed to "Most Recent"
// @compatible  chrome
// @compatible  firefox
// @compatible  opera
// @compatible  safari
// @compatible  msedge
// @license     MIT
// @icon        https://en.facebookbrand.com/wp-content/uploads/2019/04/f_logo_RGB-Hex-Blue_512.png
// @match       https://*.facebook.com/*
// @grant       none
// ==/UserScript==
(function(){ var arr = [
    "",
    "// ### Global vars ###",
    "var cur = new URL(window.location.href), spurl = cur['href'].split('/'),",
    "",
    "// ### Feed vars ###",
    "fd = cur['origin'] + '?sk=h_chr', elm = document.querySelectorAll(\"[data-click='bluebar_logo'] > a, ._3qcu > a, #navItem_4748854339 > a\"), grp = document.querySelectorAll(\"[data-type='type_group'] > a\"),  idx = ['?sk=h_nor', '?ref=logo', '?sk=nf', '?ref=tn_tnmn'],",
    "",
    "// ### Groups vars ###",
    "grid = document.querySelectorAll(\"[property='al:android:url']\"), gdisc = document.querySelectorAll('._2yau'), gpop = ['CHRONOLOGICAL', 'RECENT_ACTIVITY'], gref = ['?ref=group_header', '?ref=bookmarks', '?ref=direct', '?fref=nf', '?ref=nf_target&fref=nf'];",
    "",
    "",
    "// ### For Feed ###",
    "if(cur['href'] === cur['origin'] + '/'){",
        "window.location.replace(fd);",
    "}",
    "idx.forEach((a) => {",
        "if(cur['search'].includes(a)){",
            "window.location.replace(fd);",
        "}",
    "});",
    "elm.forEach(elm => elm.addEventListener('click', () => {",
        "window.location.replace(fd);",
    "}), false);",
    "",
    "",
    "// ### For Groups ###",
    "grp.forEach(shc => shc.addEventListener('click', () => {",
        "shc = shc.getAttribute('href');",
        "window.location.replace(groupSec(shc, gpop[0]));",
    "}, false));",
    "",
    "if(cur['href'].includes('groups') && !cur['href'].includes('permalink') && !cur['href'].includes('comment_id')){",
        "if(spurl[5] === '' || spurl[5] === null || spurl[5] === gref[0] || spurl[5] === gref[1] || spurl[5] === gref[2] || spurl[5] === gref[3] || spurl[5] === gref[4] || spurl[5] !== '?sorting_setting=' + gpop[0]){",
            "window.location.replace(sortGrp(grid, 'content', gpop[0], cur['origin'] + '/'));",
        "}",
    "}",
    "",
    "// # Discussion link event #",
    "if(gdisc[1] !== undefined){",
        "var abpg = document.querySelectorAll('._2yaa');",
        "abpg = abpg[1].getAttribute('data-key');",
        "if(!abpg.includes('tab_about')){",
            "gdisc[1].addEventListener('click', () => { window.location.replace(sortGrp(grid, 'content', gpop[0], cur['origin'] + '/')) }, false);",
        "}",
    "}",
    "",
    "// # 'post time' & 'comment time' link event #",
    "var postime = document.querySelectorAll('._5pcq');",
    "postime.forEach(ptm => ptm.addEventListener('click', () => {",
        "ptm = ptm.getAttribute('href');",
        "window.location.replace(ptm);",
    "}, false));",
    "",
    "var comtime = document.querySelectorAll('._6qw7');",
    "comtime.forEach(cmt => cmt.addEventListener('click', () => {",
        "cmt = cmt.getAttribute('href');",
        "window.location.replace(cmt);",
    "}, false));",
    "",
    "",
    "// ### Required functions ###",
    "// # Arguments: a = element where group ID is located, b = atribute where group ID is, c = type of group disposition (CHRONOLOGICAL or RECENT_ACTIVITY), d = Facebook domain, rs = a resource variable #",
    "function sortGrp(a, b, c, d, rs){",
        "a = a[0].getAttribute(b);",
        "rs = a.split('/');",
        "return d + 'groups/' + rs[3] + '/?sorting_setting=' + c;",
    "}",
    "",
    "// # Arguments: a = href attribute, b = type of group disposition (CHRONOLOGICAL or RECENT_ACTIVITY), rs = a resource variable #",
    "function groupSec(a, b, rs){",
        "rs = a.toString().replace(/(.?)ref=bookmarks/gi, \"\");",
        "return rs + \"?sorting_setting=\" + b;",
    "}",
    "",
    ].join("\n");
    var js = document.createElement("script");
    js.type = "text/javascript";
    js.appendChild(document.createTextNode(arr));
    document.querySelectorAll("head")[0].appendChild(js);
})();
